### ISSUE TITLE
Terraform Test:  terraform doesn't use impersonate GCP authentication by module test run

### ISSUE THREAD

**Issue by constin** [hashicorp/terraform#34129]
### Terraform Version

```shell
terraform v1.6.2
Use Cases:
I always use impersonate in my code. But it's not possible to use tests with it, due terraform ignore impersonate authentication for test command.
```

### Terraform Configuration Files

provider.tf
```
data "google_service_account_access_token" "default" {
  provider               = google.impersonate
  target_service_account = "test-account@some_project_id@iam.gserviceaccount.com"
  scopes                 = ["userinfo-email", "cloud-platform"]
  lifetime               = "600s"
}

provider "google" {
  alias = "impersonate"
}

provider "google" {
  access_token = data.google_service_account_access_token.default.access_token
}
```

modules/repository/main.tf
```
# example resources that need to be created
resource "google_artifact_registry_repository" "repo" {
  project       = local.project_id
  location      = "europe-west1"
  repository_id = var.registry_name
  format        = "DOCKER"
  lifecycle {
    prevent_destroy = false
  }
}

variable "registry_name" {
  type = string
}
```
main.tf
```
module "repository" {
  source     = "./modules/repository"
  registry_name = var.registry_name
}

```
tests.tftest.hcl
```

variables {
  registry_name      = "terraform-test"
}

run "test" {
  module {
    source = "./modules/repository"
  }
  assert {
    condition     = google_artifact_registry_repository.repo.name != "terraform-test"
    error_message = "repository name did not match expected"
  }
}
```
### Debug Output

```
valid_string_concat.tftest.hcl... in progress
  run "test"... fail
╷
│ Error: Error creating Repository: googleapi: Error 403: Permission 'artifactregistry.repositories.create' denied on resource '//artifactregistry.googleapis.com/projects/some_project_id/locations/europe-west1' (or it may not exist).
│ Details:
│ [
│   {
│     "@type": "type.googleapis.com/google.rpc.ErrorInfo",
│     "domain": "artifactregistry.googleapis.com",
│     "metadata": {
│       "permission": "artifactregistry.repositories.create",
│       "resource": "projects/some_project_id/locations/europe-west1"
│     },
│     "reason": "IAM_PERMISSION_DENIED"
│   }
│ ]
│ 
│   with google_artifact_registry_repository.repo,
│   on modules/repository/main.tf line 6, in resource "google_artifact_registry_repository" "repo":
│    6: resource "google_artifact_registry_repository" "repo" {
│ 
╵
tests.tftest.hcl... tearing down
tests.tftest.hcl... fail
```

### Expected Behavior

terraform must be authored under the user "test-account@some_project_id@iam.gserviceaccount.com" while running this module  test , according default provider.  all works perfect by terraform apply

### Actual Behavior

terraform authored under  my private user

### Steps to Reproduce

terraform init
terraform test

**Comment by liamcervante** [comment]
Hi @constin, thanks for filing this.

I see you also call the module from your `main.tf` configuration file. If you remove the `module` argument from your `run "test"` block does it work as expected?

For example:

```
# tests.tftest.hcl
variables {
  registry_name      = "terraform-test"
}

run "test" {
  assert {
    condition     = google_artifact_registry_repository.repo.name != "terraform-test"
    error_message = "repository name did not match expected"
  }
} 
```

`terraform test` doesn't initialise the providers in the main configuration itself, and so doesn't make them available to any module calls you make directly from the testing file. If you remove the module block from your run block, Terraform will instead just execute the main configuration where you have your module call defined anyway. In that context Terraform will use the providers available to the main configuration.

**Comment by constin** [comment]
> f you remove the `module` argument from your `run "test"` block does it work as expected?

hi @liamcervante 
yes,If I call this  test for the whole code (without "module" argument in this block), everything is ok.

**Comment by liamcervante** [comment]
Thanks for the extra context. 

Normally you could achieve this by defining the providers directly within the test file, and then providing them directly to the run block as part of the test file. Unfortunately, you are relying on an intermediary data block to initialise the provider which isn't supported yet by the testing framework. We intend to release support for this in the next minor version (`v1.7`) and you can track the progress in these tickeets:

- https://github.com/hashicorp/terraform/issues/34100
- https://github.com/hashicorp/terraform/issues/34007

In the meantime, a workaround I would suggest for testing the submodule directly would be to build a dedicated testing/wrapping module that just initialises the providers you need and calls the module you want to test outside the testing framework in a single configuration file. You could then write tests directly against that module, or use the `module` block from the configuration file as you have been doing here. With this approach you would need to build additional outputs to expose everything you wanted to check in your assertions.

Sorry this isn't supported yet, but we are working on it and you should see an improvement for this use-case in upcoming releases.

Given we have the above two tickets tracking work to support this, I will mark this as a duplicate. I'll still monitor the ticket and will respond it you have questions or follow up about the work around.

Thanks!

**Comment by liamcervante** [comment]
Actually, another alternative just jumped out to me. You can use resource targeting to target the specific module you want to test within the configuration so the test won't execute or create any other resources or modules.

```hcl
run "test" {

  plan_options {
    target = [ module.repository ]
  }

  // ... assertions ...
}
```

You might still need to make some modifications to expose exactly what you want to assert on, but you wouldn't need a wrapper module with this approach.

**Comment by constin** [comment]
Thank you for answer.
plan_options argument will start plan , in this case I need apply. 

At the moment I want to understand for myself if and how these tests can be used in production. I need a full-fledged flexible solution.  So I'll wait for 1.7 and see if it suits my needs.

Anyway thank you so much for such a detailed and quick reply. It helped me not to spend more time on this issue and get on with other tasks :)

**Comment by liamcervante** [comment]
**Call out for testing!** 

We just released a alpha for [v1.7](https://github.com/hashicorp/terraform/releases/tag/v1.7.0-alpha20231108) yesterday.

In the alpha you can reference variables and outputs from run blocks in your provider configurations. This should allow you to create a setup module that uses one provider, load credentials or configuration you need for a second provider from the first provider, and then pass the required data into the second provider.

For example:

```hcl
# testing/setup/main.tf

data "vault_generic_secret" "aws_credentials" {
  path = "some/path/to/secret"
}

output "aws_access_key_id" {
  sensitive = true
  value = data.vault_generic_secret.aws_credentials.data["AWS_ACCESS_KEY_ID"]
}

output "aws_secret_access_key" {
  sensitive = true
  value = data.vault_generic_secret.aws_credentials.data["AWS_SECRET_ACCESS_KEY"]
}
```

```hcl
# tests/main.tftest.hcl

provider "vault" {
  // ... some configuration ...
}

provider "aws" {
  region     = "us-west-2"
  access_key = run.setup.aws_access_key_id
  secret_key = run.setup.aws_secret_access_key
}

run "setup" {
  module {
    source = "./testing/setup"
  }
}

run "test" {
  // ... a normal testing block ...
}
```

I believe this will address all the use cases mentioned in this ticket. If anyone wants to try it out, please do and file and bugs or enhancement requests you have in new Github tickets. Alternatively jump over to the hashicorp forum and leave your feedback and questions in this [post](https://discuss.hashicorp.com/t/request-for-testing-reference-variable-and-run-block-outputs-in-provider-configurations-during-terraform-test/59907).

Thanks!
